Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,392 @@
+ *
+ * @return Unmutable array of font family name strings that are installed in the system.
+ */
+static NSArray<NSString*>* _dwriteGetFamilyNames() {
+    NSMutableArray<NSString*>* fontFamilyNames = [NSMutableArray<NSString*> array];
+
+    // Get the direct write factory instance
+    ComPtr<IDWriteFactory> dwriteFactory = GetDWriteFactoryInstance();
+
+    // Get the system font collection.
+    ComPtr<IDWriteFontCollection> fontCollection;
+    THROW_IF_FAILED(dwriteFactory->GetSystemFontCollection(&fontCollection));
+
+    // Get the number of font families in the collection.
+    uint32_t count = 0;
+    count = fontCollection->GetFontFamilyCount();
+
+    for (uint32_t i = 0; i < count; ++i) {
+        // Get the font family.
+        ComPtr<IDWriteFontFamily> fontFamily;
+        THROW_IF_FAILED(fontCollection->GetFontFamily(i, &fontFamily));
+
+        // Get a list of localized strings for the family name.
+        ComPtr<IDWriteLocalizedStrings> familyNames;
+        THROW_IF_FAILED(fontFamily->GetFamilyNames(&familyNames));
+
+        [fontFamilyNames addObject:_convertLocalizedStringToNSString(familyNames.Get())];
+    }
+
+    return [fontFamilyNames autorelease];
+}
+
+/**
+ * Helper method to create _CTLine objects given a _CTTypesetter, attributed string range to use and frame size to fit in.
+ *
+ * @parameter ts _CTTypesetter object to use.
+ * @parameter range attributed string range to use.
+ * @parameter frameSize size parameters of the frame to fit the text into.
+ *
+ * @return Unmutable array of _CTLine objects created with the requested parameters.
+ */
+static NSArray<_CTLine*>* _dwriteGetLines(_CTTypesetter* ts, CFRange range, CGRect frameSize) {
+    // Call custom renderer to get all glyph run details
+    _DWriteGlyphRunDetails glyphRunDetails;
+    _callCustomRender(ts, range, frameSize, &glyphRunDetails);
+
+    // Create _CTLine objects from the the obtained glyph run details
+    int numOfGlyphRuns = glyphRunDetails._glyphRuns.size();
+    int i = 0;
+    int j = 0;
+
+    NSMutableArray<_CTLine*>* lines = [NSMutableArray array];
+
+    while (j < numOfGlyphRuns) {
+        _CTLine* line = [[_CTLine new] autorelease];
+        NSMutableArray<_CTRun*>* runs = [NSMutableArray array];
+        uint32_t stringRange = 0;
+        // Glyph runs that have the same _baselineOriginY value are part of the the same Line.
+        while ((j < numOfGlyphRuns) && (glyphRunDetails._baselineOriginY[i] == glyphRunDetails._baselineOriginY[j])) {
+            j++;
+        }
+        while (i < j) {
+            // Create _CTRun objects and make them part of _CTLine
+            _CTRun* run = [[_CTRun new] autorelease];
+            run->_range.location = glyphRunDetails._glyphRunDescriptions[i]._textPosition;
+            run->_range.length = glyphRunDetails._glyphRunDescriptions[i]._stringLength;
+            run->_stringFragment = [ts->_string substringWithRange:NSMakeRangeFromCF(run->_range)];
```

Comment:
We might need the fragments for draw so I cached it in the object instead of sending a message to NS to calculate every time. Also it was very useful for debugging purposes. Will not over optimize now given we are still developing and this is not a huge memory over head

---
In reply to: [77757111](https://github.com/Microsoft/WinObjC/pull/950#discussion_r77757111) [](ancestors = 77757111)